### ISSUE TITLE
Unexpected token. Did you mean `{'>'}` or `&gt;`?

### ISSUE THREAD

**Issue by mgholam** [sveltejs/language-tools#143]
![svelte](https://user-images.githubusercontent.com/460565/83326338-7d520280-a288-11ea-8265-a5303c1bb390.png)

**Describe the bug**
On a perfectly good `app.svelte ` file I'm getting the above.

**System (please complete the following information):**
 - OS: Windows
 - IDE: VSCode
 - Plugin/Package: Svelte Beta

**Comment by dummdidumm** [comment]
Could you paste the complete code of that file?

**Comment by jasonlyu123** [comment]
This is probably because of the code before it got transformed with a missing brackets. Can you provide the code before it?

**Comment by mgholam** [comment]
```js
<script>
  import { onMount } from "svelte";
  import u from "./Helper/utils.js";
  import msgbox from "./Components/msgbox.js";
  import Tabs from "./Components/tabs.svelte";
  import Button from "./Components/Button.svelte";
  import InboxTab from "./Main/Inbox.svelte";
  import ArchiveTab from "./Main/Archive.svelte";
  import HelpTab from "./Main/Help.svelte";
  import SettingsTab from "./Main/Settings.svelte";
  import DocOutTab from "./Main/DocOut.svelte";

  let mainid = "mainid";
  let server = "";
  let TABS;
  let timer = null;
  // let RTL = false;

  onMount(() => {
    u.addPrototypes();

    TABS.addtab({
      name: "inbox",
      description: "$Inbox$",
      icon: "gi-save-file", //icons.faFolderOpen
    });
    TABS.addtab({
      name: "docout",
      description: "$Docs Out$",
      icon: "gi-open-file", //icons.faFilter
    });
    TABS.addtab({
      name: "archive",
      description: "$Search Letters$",
      icon: "gi-search", //icons.faFilter
    });
    TABS.addtab({
      name: "settings",
      description: "$Settings$",
      icon: "gi-cog", // icons.faCog
    });
    TABS.addtab({
      name: "help",
      description: "$Help$",
      icon: "gi-question-sign",
    });

    setTimeout(() => {
      // kludge to get the server for fetch
      u.GET("/api/getuser").then((r) => (server = r));

      // server = window.ServerURL;
      u.GET("/api/inbox.getsettings").then((data) => {
        if (data.Color) {
          u.SetTheme(data.Color);
        }
        if (data.Language == "fa") {
          document.body.dir = "rtl";
          // document.documentElement.style.setProperty("--tblAlign", "right");
        }
      });
      TABS.setInitial();
    }, 0);

    // Heartbeat for auto update
    timer = setInterval(() => {
      u.GET("/api/getuser").catch(() => {
        clearInterval(timer);
        msgbox.Ok("$Connection Failed!$", "", () => window.location.reload(true));
      });
    }, 10 * 1000);
  });

  // register all tabs to be created here for Tab component
  function createTab(tabname, id, props) {
    var tab = null;

    switch (tabname) {
      case "inbox":
        tab = new InboxTab({
          target: document.getElementById(id),
        });
        break;
      case "help":
        tab = new HelpTab({
          target: document.getElementById(id),
        });
        break;
      case "archive":
        tab = new ArchiveTab({
          target: document.getElementById(id),
        });
        break;
      case "settings":
        tab = new SettingsTab({
          target: document.getElementById(id),
        });
        break;
      case "docout":
        tab = new DocOutTab({
          target: document.getElementById(id),
        });
        break;
    }

    return tab;
  }

  // function configeditor() {
  //   // console.log("here");
  // }
</script>
```

**Comment by dummdidumm** [comment]
Pasting the given code into an `App.svelte` of a starter template works for me. Anything else we should know about your setup or the file?

**Comment by mgholam** [comment]
I'm not an expert, so you will have to tell me what/where to look for.

**Comment by jasonlyu123** [comment]
Can you also provide the markup and the skyle tag?

**Comment by mgholam** [comment]
I found the culprit:
```js
<svelte:head>
  <link rel="stylesheet" href="/lib/jodit.es2018.min.css" />
  <script src="/lib/jodit.es2018.min.js"> 

  </script>
</svelte:head>
```
It's the `script` tag in the `head` section that messes up the parser.

**Comment by dummdidumm** [comment]
This is another case where [our regex-parser](https://github.com/sveltejs/language-tools/blob/1177fa21ae5a1f586a4ae30748c471e2c8eafeaa/packages/language-server/src/lib/documents/utils.ts#L44) fails. Maybe we should switch to using a html parser instead [like svelte2tsx does already](https://github.com/sveltejs/language-tools/blob/1177fa21ae5a1f586a4ae30748c471e2c8eafeaa/packages/svelte2tsx/src/htmlxparser.ts#L17). There we can check if the script/styles have a parent, and if not, ignore it.

**Comment by mgholam** [comment]
I've opened another request regarding the optimization of the vs code extension size on disk (currently 150mb).
Can you comment on this?
[I have deleted the `typescript` and `prettier` folders from the `svelte-language-server\node_modules` and all seems well, but you could "build" everything and not rely on the `node_modules` right?]

**Comment by dummdidumm** [comment]
I saw this and did not have time to look into this yet. It's weekend and we all do this in our free time. Also please keep issues separate which don't have anything to do with each other.

**Comment by mgholam** [comment]
Thanks.

**Comment by sw-yx** [comment]
@dummdidumm interested in picking up this first issue. is the idea that we should use htmlxparser.ts instead of the regex parsing? are there any other alternatives you haven't discussed?

**Comment by dummdidumm** [comment]
Yes, the idea would be to make the `findVerbatimElements` function available to the language server by also exporting it through the `index.ts`, and then adjust it so that both `svelte2tsx` and `svelte-language-server` can use it. For `svelte-language-server` it would replace the inner workings of the `extractTag` function. I'm not 100% though they can share it completely, if it differs too much we might as well duplicate some of the stuff and not do the "export from `svelte2tsx`" stuff. So for example right now `findVerbatimElements` does find all script/style tags in the htmlx, but for `svelte-language-server` we only want the top level ones. There are tests on both sides for this, so you should get feedback quickly if it's working or not.

An alternative would be to somehow enhance the regex to only look for top level scripts/styles, but I feel this gets too complicated for a regex then.

**Comment by sw-yx** [comment]
ok i'm going to err on the side of duplicating it. [clean code is overrated](https://overreacted.io/goodbye-clean-code/) :)

**Comment by sw-yx** [comment]
i have set up a test for this inside of https://github.com/sveltejs/language-tools/blob/master/packages/language-server/test/lib/documents/utils.test.ts but I am not sure it is the right way to test it:

```ts
it('ignores script tag in svelte:head', () => {
            // https://github.com/sveltejs/language-tools/issues/143#issuecomment-636422045
            const text = `
            <svelte:head>
                <link rel="stylesheet" href="/lib/jodit.es2018.min.css" />
                <script src="/lib/jodit.es2018.min.js"> 
                </script>
            </svelte:head>
            <p>jo</p>
            <script>top level script</script>
            <h1>Hello, world!</h1>
            <style>.bla {}</style>
            `
            assert.deepStrictEqual(extractTag(text, 'script'), {
                // ????
                // content: 'top level script',
                // attributes: {},
                // start: 43,
                // end: 59,
                // startPos: Position.create(2, 20),
                // endPos: Position.create(2, 36),
                // container: { start: 35, end: 68 },
            });
        })
```

it doesnt give the Unexpected token error but it does match the wrong `<script>` tag here. is there a better way to test

**Comment by sw-yx** [comment]
for whoever sees this in future - i ended up completely not duplicating it at all - the requirements for `svelte language server` (top level tags only) are very different from `svelte2tsx`.

**Comment by mgholam** [comment]
The problem still exists in `99.0.39`

**Comment by dummdidumm** [comment]
Problem in `svelte2tsx`, output has a syntax error:

```jsx
// ....
<sveltehead>
  <link rel="stylesheet" href="/lib/jodit.es2018.min.css" />
  <script src=""/lib/jodit.es2018.min.js"">  // <<<-- two `"` each, syntax error

  </script>
</sveltehead>
```

**Comment by mgholam** [comment]
The problem still exists in `99.0.41`

**Comment by dummdidumm** [comment]
I cannot reproduce the bug with the latest version, given your code snippet with `<svelte:head>..`.

**Comment by mgholam** [comment]
![sv](https://user-images.githubusercontent.com/460565/84233267-3fb26c80-ab07-11ea-96e6-50cf93a72382.png)

**Comment by dummdidumm** [comment]
Could you post a minimum snippet that reproduces the error for you?

**Comment by mgholam** [comment]
```js
<script>
  import { onMount } from "svelte";
  let aa = false;
  onMount(() => {  // <--- the problem seems to be here at =>
    aa = true;
  });
</script>

<svelte:head>
  <link rel="stylesheet" href="/lib/jodit.es2018.min.css" />
  <script src="/lib/jodit.es2018.min.js">

  </script>
</svelte:head>
```

**Comment by dummdidumm** [comment]
Can reproduce with this, thanks.

**Comment by dummdidumm** [comment]
The problem is that `svelte2tsx` will set any script it comes across as the primary script instance, so last one wins ([relevant function](https://github.com/sveltejs/language-tools/blob/429a1647531558bb271495a782dd03db70a8160e/packages/svelte2tsx/src/svelte2tsx.ts#L196)). Nesting is not taken into account.

My first attempt at fixing that was to take into account the parent node of the walk, but that does not work because parent is not always the expect parent.

To fix that I tried to extract the logic @sw-yx has implemented into `svelte2tsx` so a common ground can be used by both `svelte2tsx` and `language-server`. This works, but I then stumbled across another problem: `parse5` is very html spec compliant, also when it comes to self-closing tags. That means that something like `<SelfClosing /> <p>bla</p>` will not parse as we would expect - the node would also span `<p>bla</p>` and more, until `parse5` would see a real closing tag (`</SelfClosing>`).

In short: We have to find another way than `parse5` to extract our top level scripts.

Short Test that fails with the current implementation:

```ts

        it('can extract with self-closing component before it', () => {
            const extracted = extractTag('<SelfClosing /><style></style>', 'style');
            assert.equal(!!extracted, true);
        });
```

**Comment by mgholam** [comment]
Tried this as a workaround but it fails also:
```js
<script>
  import { onMount } from "svelte";
  let aa = false;
  onMount(mount);

  function mount() {
    aa = true;   // <---error here
  }
</script>

<svelte:head>
  <script src="/lib/jodit.es2018.min.js">

  </script>
</svelte:head>
```

**Comment by dummdidumm** [comment]
As a workaround you need your top level script after the script in `svelte:head`. Last script wins and is treated as the top level script.

**Comment by mgholam** [comment]
Can you do what the svelte compilers does (just asking)?

**Comment by dummdidumm** [comment]
This error has not much to do with the svelte compiler. It's an error which occurs while doing a transformation from svelte code to jsx code using `svelte2tsx`.

**Comment by mgholam** [comment]
> As a workaround you need your top level script after the script in svelte:head. Last script wins and is treated as the top level script.

Can you make the first script "win" and be the top level script? (as a thought...)

**Comment by dummdidumm** [comment]
Yes this could be a good quickfix in the meantime. I also have another idea, will have to test it out.

**Comment by dummdidumm** [comment]
Third time's a charm - let's hope it's fixed with the next deployment in about 12 hours.

I created a separate issue for the self-closing-tag-bug (#194)

**Comment by mgholam** [comment]
Woo Hoo! Thanks!

**Comment by mgholam** [comment]
Regression if the script block is commented : 
```
<svelte:head>
    <!-- <script src="/lib/jodit.es2018.min.js">
  </script>-->
</svelte:head>
```

**Comment by dummdidumm** [comment]
Cannot reproduce with your snippet.

**Comment by mgholam** [comment]
```
<script type="ts">
  import { onMount } from "svelte";
  // let aa = false;
  // let pp: string = "";
  // onMount(mount);

  // function mount() {
  //   aa = true;
  // }

  // function click() {
  //   pp = "ppppp";
  //   console.log("hello");
  // }
</script>

<svelte:head>
  <!-- <script src="/lib/jodit.es2018.min.js">

  </script> -->
</svelte:head>

<div>
  <button on:click={click}>hello</button>
  <input bind:value={pp} />

</div>
```

**Comment by dummdidumm** [comment]
Fix available soon.

**Comment by mgholam** [comment]
Still exists in `101.9.3` 
```js
<script >
  import { onMount } from "svelte";

  // } // <- error here

</script>

<svelte:head>
  <!-- <script src="/lib/jodit.es2018.min.js">

  </script> -->
</svelte:head>
```

**Comment by azhar1598** [comment]
I have encountered this issue while i was developing my code in VScode editor.And,there is nothing wrong with the code and there are no problems showing in the terminal and Program is also executing  very well.And then i just restarted my vscode editor ,everything got resolved I mean the error which was quoted in the question.
Just try restarting your editor.Maybe,it might get resolved.

**Comment by jasonlyu123** [comment]
Probably not related. The codebase has changed a lot since then. If you have a `>` character as text in the template there is a PR right now. Else you can provide a reproducible in a new issue.

**Comment by unclebay143** [comment]
For anyone facing this issue: I uninstalled `prettier`, restart VScode, and re-install `prettier` and everything works fine. This is the second time this happened and that same step resolved it.

**Comment by lohitakshsingla0** [comment]
I am also facing the same issue, my code is below
import React from "react";
```
const SortCards = ({listData}) => {

    return (
      <div>
        <select>
        listData.map((result) => (<option>{result.strCategory}</option>))
      </select>
      </div>
    )
  
    
}

export default SortCards;
```

Error: 
Unexpected token. Did you mean `{'>'}` or `&gt;`?ts(1382)

**Comment by lohitakshsingla0** [comment]
Even I am facing the same issue:

`
const SortCards = ({listData}) => {

    return (
      <div>
        <select>
        listData.map((result) => (<option>{result.strCategory}</option>))
      </select>
      </div>
    )
  
    
}

export default SortCards;
`

**Comment by lohitakshsingla0** [comment]
Even I am facing the same issue:

`
const SortCards = ({listData}) => {

    return (
      <div>
        <select>
        listData.map((result) => (<option>{result.strCategory}</option>))
      </select>
      </div>
    )
  
    
}

export default SortCards;
`

**Comment by lohitakshsingla0** [comment]
> For anyone facing this issue: I uninstalled `prettier`, restart VScode, and re-install `prettier` and everything works fine. This is the second time this happened and that same step resolved it.

I tried this, but doesn't worked for me

**Comment by dummdidumm** [comment]
Your code snippet is JSX code, Svelte does not support JSX.
